UnleashClient/__init__.py

```diff
@@ -450,69 +432,51 @@ def get_variant(self, feature_name: str, context: Optional[dict] = None) -> dict
-                        LOGGER.log(
-                            self.unleash_verbose_log_level,
-                            "Error in event callback: %s",
-                            excep,
-                        )
-                        return variant_check
-
-                return variant_check
-            except Exception as excep:
+        variant = self._resolve_variant(feature_name, context)
+
+        if not variant:
+            if self.unleash_bootstrapped or self.is_initialized:
                 LOGGER.log(
                     self.unleash_verbose_log_level,
-                    "Returning default flag/variation for feature: %s",
+                    "Attempted to get feature flag/variation %s, but client wasn't initialized!",
                     feature_name,
                 )
+            variant = DISABLED_VARIATION
+
+        self.engine.count_variant(feature_name, variant["name"])
+        self.engine.count_toggle(feature_name, variant["feature_enabled"])
+
+        if self.unleash_event_callback and self.engine.should_emit_impression_event(
+            feature_name
+        ):
+            try:
+                event = UnleashEvent(
+                    event_type=UnleashEventType.VARIANT,
+                    event_id=uuid.uuid4(),
+                    context=context,
+                    enabled=variant["enabled"],
+                    feature_name=feature_name,
+                    variant=variant["name"],
+                )
+
+                self.unleash_event_callback(event)
+            except Exception as excep:
                 LOGGER.log(
                     self.unleash_verbose_log_level,
-                    "Error checking feature flag variant: %s",
+                    "Error in event callback: %s",
                     excep,
                 )
 
-                # The feature doesn't exist, so create it to track metrics
-                new_feature = Feature.metrics_only_feature(feature_name)
-                self.features[feature_name] = new_feature
+        return variant
 
-                # Use the feature's get_variant method to count the call
-                variant_check = new_feature.get_variant(context)
-                return variant_check
-        else:
-            LOGGER.log(
-                self.unleash_verbose_log_level,
-                "Returning default flag/variation for feature: %s",
-                feature_name,
-            )
-            LOGGER.log(
-                self.unleash_verbose_log_level,
-                "Attempted to get feature flag/variation %s, but client wasn't initialized!",
-                feature_name,
-            )
-            return DISABLED_VARIATION
+    def _resolve_variant(self, feature_name: str, context: dict) -> dict:
+        """
+        Resolves a feature variant.
+        """
+        variant = self.engine.get_variant(feature_name, context)
+        if variant:
+            return {k: v for k, v in asdict(variant).items() if v is not None}
```

Comment:
Just curious: Why do we need this?

Reply:
It's a good question. Yggdrasil core returns a dataclass here (which is like an object without methods, similar to a Rust struct). I'm a fan of dataclasses because they're quite explicit about their intent. The SDK at the moment returns an untyped dict. So I don't want to change it for 3 reasons:

- Yggdrasil abstractions should stay in Yggdrasil, bleeding internal classes outside of the domain boundary is poor encapsulation
- This is one less thing for consumers to have to think about or worry about, that one is quite important to me in Python because its quite hard to know all the places you may be using a return type. This means no one has to hunt down all their variants and change the way they access data on them
- Dataclasses do optionals, which become nulls. The client spec suggests that variants without payloads shouldn't include that field rather than including a null. The dict comprehension here is filtering out nulls